IOW-728 (Notify a human and refuse to create an empty RDB file)

orosz-usgs opened the pull request:
Exception thrown in processRequest.processRequest() if the
created RDB file has zero data rows.
    Added test BuildRdbFileTest.testEmptyRdbException()

Before making a pull request
----------------------------

- [ ] Make sure all tests run
- [ ] Update the changelog appropriately

Title
-----------
Brief description of changes. Reference the JIRA ticket if appropriate

Description
-----------
If no JIRA ticket is referenced, describe the changes made. Note anything that you want the reviewers to know while
reviewing your pull request

After making a pull request
---------------------------
- [ ] If appropriate, put the link to the PR in the JIRA ticket
- [ ] Assign someone to review unless the change is trivial

orosz-usgs committed:
IOW-728 (Notify a human and refuse to create an empty RDB file)
    Exception thrown in processRequest.processRequest() if the
created RDB file has zero data rows.
    Added test BuildRdbFileTest.testEmptyRdbException()

ee-usgs commented:
Ed and I talked a bit about this MR:
I was thinking of an empty RDB file as a WARNING level situation:  No errors were thrown by the code, we know how to continue, but we violated a business rule, that is, _Do not generate an empty RDB file_.  In the world where we consider this a business rule violation, we we notify a human of a warning level event and terminate normally.

In this particular situation, the difference between a warning and an error is arbitrary:  This is the end of the pipeline and if we consider an empty file an error, it would result in a notification being sent (although the pipeline would do retries).  So, this MR is workable as is.

In my broader scheme, we would have an SNS to accept WARNING level messages and have the ability to send validation messages to it at the row level (ie, not throw an error, just send a message).  This would allow us to send validation warnings like "I see an approval code I don't recognize, but I can continue anyway", let a human get alerts from a pipeline-level SNS for this type of thing, and continue on.

duselmann reviewed (approved):
If there is an odd/one-off need to send an empty file to delete all data then a blank file could be manually placed in the S3 Bucket. I do not suspect we will need to send empty files in production.

ssoper-usgs commented on src/main/java/gov/usgs/wma/waterdata/groundwater/BuildRdbFile.java:
```diff
@@ -88,12 +88,19 @@ protected ResultObject processRequest(String locationFolder) {
 			RdbWriter rdbWriter = createRdbWriter(writer).writeHeader();
 			dao.sendDiscreteGroundWater(states, rdbWriter, aqDao.getParameters());
 
+			if (rdbWriter.getDataRowCount() == 0) {
```
Can we add a comment just above this if block briefly explaining why we don't want to send the file to s3?

ee-usgs replied:
I think a separate comment wouldn't be needed if the RuntimeException was more descriptive.  What if the RE mentions that this is a "rule violation as described in JIRA IOW-728".... Or something similar.

ssoper-usgs commented on src/test/java/gov/usgs/wma/waterdata/groundwater/BuildRdbFileTest.java:
```diff
@@ -215,6 +217,64 @@ protected RdbWriter createRdbWriter(Writer destination) {
+
+		// ASSERTIONS
+		Mockito.verify(mockAqDao, Mockito.atLeastOnce()).getParameters();
+		Mockito.verify(mockAqDao, Mockito.atMostOnce()).getParameters();
```
Is there a method on `Mockito` like `once()` or something similar? It seems like we're testing that something about these parameters happened exactly 1 time, but I could be totally wrong.

duselmann replied:
I think there is a times(1) that would achieve that.

ssoper-usgs commented:
One thing that came up in our conversation to keep in mind is that this code is not part of the state machine.  Rather, it is executed on a schedule defined in a cron in the `serverless.yml` file.

ee-usgs reviewed (approved):
Based on the conversation between Ed, Sam and myself, I think it is OK to merge this change as is.  Per our discussion, we can later replace the error with an SNS notification.
Alternately, if Ed wants to morph this MR into the SNS approach, that would be OK too.

orosz-usgs committed:
IOW-728 (Added notify human)
    Added aws-java-sdk-sqs to maven dependencies
    New util class: SqsUtil
    Updated serverless.yml to create SQS log
queue.
    Updated BuildRdbFile to send a info and error
messages to SQS using SqsUtil
    Udated tests to check for addSQSMessage being
called.

orosz-usgs committed:
IOW-728 Added dev to environments

orosz-usgs committed:
IOW-728 Set queue timeout

orosz-usgs committed:
IOW-728  Fixed queue timeout

orosz-usgs committed:
IOW-728  Removed DEV environment. was incomplete & not needed
for the time being.

orosz-usgs committed:
IOW-728 added logging to no state found error.

orosz-usgs committed:
Merge pull request #1 from orosz-usgs/IOW-728-notify

Iow 728 notify

duselmann reviewed (approved):
Looks good to me.

ayan-usgs commented on serverless.yml:
```diff
@@ -119,6 +121,12 @@ resources:
         EvaluationPeriods: 1
         AlarmActions:
           - Ref: snsTopic
+    logQueue:
+      Type: AWS::SQS::Queue
+      Properties:
+        QueueName: aqts-capture-etl-rdb-log-${self:provider.stage}
+        MessageRetentionPeriod: 1209600
+        VisibilityTimeout : 180
```
What's the purpose of this log queue?

ssoper-usgs replied:
I'm similarly confused - do we want to use an SNS or SES topic instead?  As written, I think this would populate an SQS queue with log information, but nothing would be done with it.

orosz-usgs committed:
IOW-728  Cleanup per review. Added comment as to why an
empty RDB file is not allowed.
  Refactored tests to use Mockito.times(1)

orosz-usgs committed:
Merge branch 'master' into IOW-728

orosz-usgs committed:
Merge branch 'master' into IOW-728

orosz-usgs committed:
IOW-728  Removed the notification logic. The sns notifications
will be added later.

orosz-usgs committed:
IOW-728  Removed the notification logic. The sns notifications
will be added later.

kkehl-usgs commented on src/main/java/gov/usgs/wma/waterdata/groundwater/BuildRdbFile.java:
```diff
@@ -75,25 +75,40 @@ protected ResultObject processRequest(String locationFolder) {
 			result.setMessage("Count is rows written to file: " + s3bucket.getKeyName());
+			mess = String.format("INFO: RDB file created, %d rows %s",
+					rdbWriter.getDataRowCount(), details);
+			System.out.println(mess);
```
We want to use system.out.println() and not a logger?

orosz-usgs replied:
Using a logger is a todo. It will still go to cloud watch.

ssoper-usgs reviewed (approved):
Looks good, just that one system.out.println to take care of.
